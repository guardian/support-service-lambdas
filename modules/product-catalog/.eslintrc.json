{
  "plugins": ["sort-keys-fix"],
  "overrides": [
    {
      "files": ["src/productCatalogSchema.ts", "src/productBillingPeriods.ts"],
      "rules": {
        "sort-keys-fix/sort-keys-fix": "warn"
      }
    }
  ]
}