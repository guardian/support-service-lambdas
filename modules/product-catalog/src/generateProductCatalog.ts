import { arrayToObject } from '@modules/arrayFunctions';
import { isBillingPeriod } from '@modules/billingPeriod';
import type {
	ZuoraCatalog,
	ZuoraProductRatePlan,
	ZuoraProductRatePlanCharge,
} from '@modules/zuora-catalog/zuoraCatalogSchema';
import type { ProductCatalog } from '@modules/product-catalog/newProductCatalogTypes';
import { oneTimeContribution } from '@modules/product-catalog/oneTimeContributionProduct';
import { stripeProducts } from '@modules/product-catalog/stripeProducts';
import {
	activeProducts,
	getProductRatePlanChargeKey,
	getProductRatePlanKey,
	getZuoraProductKey,
	isSupportedProduct,
	isSupportedProductRatePlan,
} from '@modules/product-catalog/zuoraToProductNameMappings';

type NonNullPrice = { currency: string; price: number };
type PricingObject = Record<string, number>;

const roundPriceToTwoDecimalPlaces = (price: number) => {
	return parseFloat(price.toFixed(2));
};
const getPricingObject = (charges: ZuoraProductRatePlanCharge[]) => {
	const allPrices = charges.flatMap((charge) => {
		return charge.pricing
			.map(({ currency, price }) => ({ currency, price }))
			.filter<NonNullPrice>(
				(price): price is NonNullPrice => price.price != null,
			);
	});

	return allPrices.reduce<PricingObject>((acc, price) => {
		const total = (acc[price.currency] ?? 0) + price.price;
		acc[price.currency] = roundPriceToTwoDecimalPlaces(total);
		return acc;
	}, {});
};
const getProductRatePlanCharges = (
	productRatePlanCharges: ZuoraProductRatePlanCharge[],
) => {
	return arrayToObject(
		productRatePlanCharges.map((productRatePlanCharge) => {
			const productRatePlanChargeName = getProductRatePlanChargeKey(
				productRatePlanCharge.name,
			);
			return {
				[productRatePlanChargeName]: {
					id: productRatePlanCharge.id,
				},
			};
		}),
	);
};

/**
 * We use this to hoist billing period from the `productRatePlanCharges` up into the `ratePlans`
 * if-and-only-if all the charges have the same `billingPeriod`.
 */
const getBillingPeriod = (productRatePlan: ZuoraProductRatePlan) => {
	const billingPeriods = new Set(
		productRatePlan.productRatePlanCharges.map(
			(productRatePlanCharge) => productRatePlanCharge.billingPeriod,
		),
	);
	if (billingPeriods.size > 1) {
		const errorList = [...billingPeriods].join(',');
		throw new Error(
			`Product rate plan ${productRatePlan.name} has multiple billingPeriods ${errorList}`,
		);
	}

	const [billingPeriod] = billingPeriods;
	// billingPeriod as null is valid from Zuora, which is used for one-time charges e.g. gifts
	if (billingPeriod === null) {
		return;
	}

	if (!isBillingPeriod(billingPeriod)) {
		throw new ReferenceError(`Unexpected billing period ${billingPeriod}`);
	}

	return billingPeriod;
};

const getZuoraProduct = (
	isActive: boolean,
	productRatePlans: ZuoraProductRatePlan[],
) => {
	return {
		billingSystem: 'zuora',
		active: isActive,
		ratePlans: arrayToObject(
			productRatePlans
				.filter((productRatePlan) =>
					isSupportedProductRatePlan(productRatePlan.name),
				)
				.map((productRatePlan) => {
					const billingPeriod = getBillingPeriod(productRatePlan);
					const productRatePlanKey = getProductRatePlanKey(
						productRatePlan.name,
					);
					return {
						[productRatePlanKey]: {
							id: productRatePlan.id,
							pricing: getPricingObject(productRatePlan.productRatePlanCharges),
							charges: getProductRatePlanCharges(
								productRatePlan.productRatePlanCharges,
							),
							...(billingPeriod && { billingPeriod }),
						},
					};
				}),
		),
	};
};
export const generateProductCatalog = (
	catalog: ZuoraCatalog,
): ProductCatalog => {
	const supportedProducts = catalog.products.filter((product) =>
		isSupportedProduct(product.name),
	);

	const result = arrayToObject(
		supportedProducts.map((product) => {
			const productName = getZuoraProductKey(product.name);
			return {
				[productName]: getZuoraProduct(
					activeProducts.includes(productName),
					product.productRatePlans,
				),
			};
		}),
	);
	const productCatalog = {
		...oneTimeContribution,
		...stripeProducts,
		...result,
	};
	return productCatalog as ProductCatalog;
};
