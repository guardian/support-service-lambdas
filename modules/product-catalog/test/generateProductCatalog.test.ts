import { zuoraCatalogSchema } from '@modules/zuora-catalog/zuoraCatalogSchema';
import { generateProductCatalog } from '@modules/product-catalog/generateProductCatalog';
import { generateTypeObject } from '@modules/product-catalog/generateTypeObject';
import code from '../../zuora-catalog/test/fixtures/catalog-code.json';
import prod from '../../zuora-catalog/test/fixtures/catalog-prod.json';

describe('prod', () => {
	test('Generated product catalog matches snapshot', () => {
		const prodZuoraCatalog = zuoraCatalogSchema.parse(prod);
		const prodProductCatalog = generateProductCatalog(prodZuoraCatalog);
		console.log(JSON.stringify(prodProductCatalog));
		expect(prodProductCatalog).toMatchSnapshot();
	});

	test('Generated product catalog types match snapshot', () => {
		const prodZuoraCatalog = zuoraCatalogSchema.parse(prod);
		const prodTypeObject = generateTypeObject(prodZuoraCatalog);
		expect(prodTypeObject).toMatchSnapshot();
	});
});

describe('code', () => {
	test('Generated product catalog matches snapshot', () => {
		const codeZuoraCatalog = zuoraCatalogSchema.parse(code);
		const codeProductCatalog = generateProductCatalog(codeZuoraCatalog);
		console.log(JSON.stringify(codeProductCatalog));
		expect(codeProductCatalog).toMatchSnapshot();
	});

	test('Generated product catalog types match snapshot', () => {
		const codeZuoraCatalog = zuoraCatalogSchema.parse(code);
		const codeTypeObject = generateTypeObject(codeZuoraCatalog);
		expect(codeTypeObject).toMatchSnapshot();
	});
});
