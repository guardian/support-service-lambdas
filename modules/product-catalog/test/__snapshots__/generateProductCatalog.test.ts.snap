// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generated product catalog matches snapshot 1`] = `
{
  "Contribution": {
    "billingSystem": "zuora",
    "ratePlans": {
      "Annual": {
        "billingPeriod": "Annual",
        "charges": {
          "Contribution": {
            "id": "2c92a0fc5e1dc084015e37f58c7b0f34",
          },
        },
        "id": "2c92a0fc5e1dc084015e37f58c200eea",
        "pricing": {
          "AUD": 100,
          "CAD": 5,
          "EUR": 60,
          "GBP": 60,
          "NZD": 60,
          "USD": 60,
        },
      },
      "Monthly": {
        "billingPeriod": "Month",
        "charges": {
          "Contribution": {
            "id": "2c92a0fc5aacfadd015ad250bf2c6d38",
          },
        },
        "id": "2c92a0fc5aacfadd015ad24db4ff5e97",
        "pricing": {
          "AUD": 10,
          "CAD": 5,
          "EUR": 5,
          "GBP": 5,
          "NZD": 5,
          "USD": 5,
        },
      },
    },
  },
  "DigitalSubscription": {
    "billingSystem": "zuora",
    "ratePlans": {
      "Annual": {
        "billingPeriod": "Annual",
        "charges": {
          "Subscription": {
            "id": "2c92a0fb4edd70c9014edeaa5001218c",
          },
        },
        "id": "2c92a0fb4edd70c8014edeaa4e972204",
        "pricing": {
          "AUD": 269.99,
          "CAD": 274,
          "EUR": 187,
          "GBP": 149,
          "NZD": 269.99,
          "USD": 249,
        },
      },
      "Monthly": {
        "billingPeriod": "Month",
        "charges": {
          "Subscription": {
            "id": "2c92a0fb4edd70c9014edeaa50342192",
          },
        },
        "id": "2c92a0fb4edd70c8014edeaa4eae220a",
        "pricing": {
          "AUD": 26.99,
          "CAD": 27.44,
          "EUR": 18.99,
          "GBP": 14.99,
          "NZD": 26.99,
          "USD": 24.99,
        },
      },
      "OneYearGift": {
        "charges": {
          "Subscription": {
            "id": "2c92a011779932fd0177a670f43102aa",
          },
        },
        "id": "2c92a00c77992ba70177a6596f710265",
        "pricing": {
          "AUD": 175,
          "CAD": 175,
          "EUR": 125,
          "GBP": 99,
          "NZD": 175,
          "USD": 165,
        },
      },
      "ThreeMonthGift": {
        "charges": {
          "Subscription": {
            "id": "2c92a00f779933030177a65881490325",
          },
        },
        "id": "2c92a00d779932ef0177a65430d30ac1",
        "pricing": {
          "AUD": 63,
          "CAD": 63,
          "EUR": 45,
          "GBP": 36,
          "NZD": 63,
          "USD": 60,
        },
      },
    },
  },
  "GuardianLight": {
    "billingSystem": "zuora",
    "ratePlans": {
      "Monthly": {
        "billingPeriod": "Month",
        "charges": {
          "GuardianLight": {
            "id": "8a129b7892c35f170192dd3225572b97",
          },
        },
        "id": "8a12831492c341730192dd1c39207038",
        "pricing": {
          "GBP": 2,
        },
      },
    },
  },
  "GuardianPatron": {
    "billingSystem": "stripe",
    "ratePlans": {
      "GuardianPatron": {
        "billingPeriod": "Month",
        "charges": {
          "GuardianPatron": {
            "id": "guardian_patron",
          },
        },
        "id": "guardian_patron",
        "pricing": {
          "AUD": 0,
          "CAD": 0,
          "EUR": 0,
          "GBP": 0,
          "NZD": 0,
          "USD": 0,
        },
      },
    },
  },
  "GuardianWeeklyDomestic": {
    "billingSystem": "zuora",
    "ratePlans": {
      "Annual": {
        "billingPeriod": "Annual",
        "charges": {
          "Subscription": {
            "id": "2c92a0fe6619b4b901661aa8e6811695",
          },
        },
        "id": "2c92a0fe6619b4b901661aa8e66c1692",
        "pricing": {
          "AUD": 480,
          "CAD": 396,
          "EUR": 318,
          "GBP": 180,
          "NZD": 600,
          "USD": 360,
        },
      },
      "Monthly": {
        "billingPeriod": "Month",
        "charges": {
          "Subscription": {
            "id": "2c92a0fd79ac64b00179ae3f9d704962",
          },
        },
        "id": "2c92a0fd79ac64b00179ae3f9d474960",
        "pricing": {
          "AUD": 40,
          "CAD": 33,
          "EUR": 26.5,
          "GBP": 15,
          "NZD": 50,
          "USD": 30,
        },
      },
      "OneYearGift": {
        "billingPeriod": "Annual",
        "charges": {
          "Subscription": {
            "id": "2c92a0ff67cebd0d0167f0a1a85b2350",
          },
        },
        "id": "2c92a0ff67cebd0d0167f0a1a834234e",
        "pricing": {
          "AUD": 480,
          "CAD": 396,
          "EUR": 318,
          "GBP": 180,
          "NZD": 600,
          "USD": 360,
        },
      },
      "Quarterly": {
        "billingPeriod": "Quarter",
        "charges": {
          "Subscription": {
            "id": "2c92a0fe6619b4b601661aa8b74e623f",
          },
        },
        "id": "2c92a0fe6619b4b301661aa494392ee2",
        "pricing": {
          "AUD": 120,
          "CAD": 99,
          "EUR": 79.5,
          "GBP": 45,
          "NZD": 150,
          "USD": 90,
        },
      },
      "ThreeMonthGift": {
        "billingPeriod": "Quarter",
        "charges": {
          "Subscription": {
            "id": "2c92a00e6dd988e2016df853875d74c6",
          },
        },
        "id": "2c92a00e6dd988e2016df85387417498",
        "pricing": {
          "AUD": 120,
          "CAD": 99,
          "EUR": 79.5,
          "GBP": 45,
          "NZD": 150,
          "USD": 90,
        },
      },
    },
  },
  "GuardianWeeklyRestOfWorld": {
    "billingSystem": "zuora",
    "ratePlans": {
      "Annual": {
        "billingPeriod": "Annual",
        "charges": {
          "Subscription": {
            "id": "2c92a0fe6619b4b601661ab3002f2653",
          },
        },
        "id": "2c92a0fe6619b4b601661ab300222651",
        "pricing": {
          "AUD": 424,
          "CAD": 345,
          "EUR": 270,
          "GBP": 297.6,
          "NZD": 530,
          "USD": 396,
        },
      },
      "Monthly": {
        "billingPeriod": "Month",
        "charges": {
          "Monthly": {
            "id": "2c92a0ff79ac64e30179ae4566cb3a86",
          },
        },
        "id": "2c92a0ff79ac64e30179ae45669b3a83",
        "pricing": {
          "GBP": 24.8,
          "USD": 33,
        },
      },
      "OneYearGift": {
        "billingPeriod": "Annual",
        "charges": {
          "Subscription": {
            "id": "2c92a0ff67cebd140167f0a2f68912ed",
          },
        },
        "id": "2c92a0ff67cebd140167f0a2f66a12eb",
        "pricing": {
          "AUD": 424,
          "CAD": 345,
          "EUR": 270,
          "GBP": 297.6,
          "NZD": 530,
          "USD": 396,
        },
      },
      "Quarterly": {
        "billingPeriod": "Quarter",
        "charges": {
          "Subscription": {
            "id": "2c92a0ff6619bf8b01661ab2d0396eb2",
          },
        },
        "id": "2c92a0086619bf8901661ab02752722f",
        "pricing": {
          "AUD": 106,
          "CAD": 86.25,
          "EUR": 67.5,
          "GBP": 74.4,
          "NZD": 132.5,
          "USD": 99,
        },
      },
      "ThreeMonthGift": {
        "billingPeriod": "Quarter",
        "charges": {
          "Subscription": {
            "id": "2c92a0076dd9892e016df8503e936c4a",
          },
        },
        "id": "2c92a0076dd9892e016df8503e7c6c48",
        "pricing": {
          "AUD": 106,
          "CAD": 86.25,
          "EUR": 67.5,
          "GBP": 74.4,
          "NZD": 132.5,
          "USD": 99,
        },
      },
    },
  },
  "HomeDelivery": {
    "billingSystem": "zuora",
    "ratePlans": {
      "Everyday": {
        "billingPeriod": "Month",
        "charges": {
          "Friday": {
            "id": "2c92a0fd560d13880156136b73770f1e",
          },
          "Monday": {
            "id": "2c92a0fd560d13880156136b74340f36",
          },
          "Saturday": {
            "id": "2c92a0fd560d13880156136b74b80f47",
          },
          "Sunday": {
            "id": "2c92a0fd560d13230156137061435de7",
          },
          "Thursday": {
            "id": "2c92a0fd560d13880156136b73b50f26",
          },
          "Tuesday": {
            "id": "2c92a0fd560d13880156136b74780f3f",
          },
          "Wednesday": {
            "id": "2c92a0fd560d13880156136b730d0f0e",
          },
        },
        "id": "2c92a0fd560d13880156136b72e50f0c",
        "pricing": {
          "GBP": 83.99,
        },
      },
      "Saturday": {
        "billingPeriod": "Month",
        "charges": {
          "Saturday": {
            "id": "2c92a0fd5e1dcf0d015e3cb39d207ddf",
          },
        },
        "id": "2c92a0fd5e1dcf0d015e3cb39d0a7ddb",
        "pricing": {
          "GBP": 20.99,
        },
      },
      "Sixday": {
        "billingPeriod": "Month",
        "charges": {
          "Friday": {
            "id": "2c92a0ff560d311b0156136f2b50531f",
          },
          "Monday": {
            "id": "2c92a0ff560d311b0156136f2bc2532f",
          },
          "Saturday": {
            "id": "2c92a0ff560d311b0156136f2c43533f",
          },
          "Thursday": {
            "id": "2c92a0ff560d311b0156136f2b8c5327",
          },
          "Tuesday": {
            "id": "2c92a0ff560d311b0156136f2c015337",
          },
          "Wednesday": {
            "id": "2c92a0ff560d311b0156136f2b185317",
          },
        },
        "id": "2c92a0ff560d311b0156136f2afe5315",
        "pricing": {
          "GBP": 73.99,
        },
      },
      "Sunday": {
        "billingPeriod": "Month",
        "charges": {
          "Sunday": {
            "id": "2c92a0ff5af9b657015b0fea5bb83fa8",
          },
        },
        "id": "2c92a0ff5af9b657015b0fea5b653f81",
        "pricing": {
          "GBP": 20.99,
        },
      },
      "Weekend": {
        "billingPeriod": "Month",
        "charges": {
          "Saturday": {
            "id": "2c92a0fd5614305c01561dc88fb875d0",
          },
          "Sunday": {
            "id": "2c92a0fd5614305c01561dc88f8975c8",
          },
        },
        "id": "2c92a0fd5614305c01561dc88f3275be",
        "pricing": {
          "GBP": 33.99,
        },
      },
    },
  },
  "NationalDelivery": {
    "billingSystem": "zuora",
    "ratePlans": {
      "Everyday": {
        "billingPeriod": "Month",
        "charges": {
          "Friday": {
            "id": "8a12999f8a268c57018a27ebe44c14b6",
          },
          "Monday": {
            "id": "8a12999f8a268c57018a27ebe55814ca",
          },
          "Saturday": {
            "id": "8a12999f8a268c57018a27ebe65914de",
          },
          "Sunday": {
            "id": "8a12999f8a268c57018a27ebe35114a6",
          },
          "Thursday": {
            "id": "8a12999f8a268c57018a27ebe4d414bf",
          },
          "Tuesday": {
            "id": "8a12999f8a268c57018a27ebe5d814d6",
          },
          "Wednesday": {
            "id": "8a12999f8a268c57018a27ebe3ce14ae",
          },
        },
        "id": "8a12999f8a268c57018a27ebe31414a4",
        "pricing": {
          "GBP": 83.99,
        },
      },
      "Sixday": {
        "billingPeriod": "Month",
        "charges": {
          "Friday": {
            "id": "8a12999f8a268c57018a27ebfecb18c7",
          },
          "Monday": {
            "id": "8a12999f8a268c57018a27ec00b418d9",
          },
          "Saturday": {
            "id": "8a12999f8a268c57018a27ec029418ea",
          },
          "Thursday": {
            "id": "8a12999f8a268c57018a27ebffb518cf",
          },
          "Tuesday": {
            "id": "8a12999f8a268c57018a27ec01a918e2",
          },
          "Wednesday": {
            "id": "8a12999f8a268c57018a27ebfde818ab",
          },
        },
        "id": "8a12999f8a268c57018a27ebfd721883",
        "pricing": {
          "GBP": 73.99,
        },
      },
      "Weekend": {
        "billingPeriod": "Month",
        "charges": {
          "Saturday": {
            "id": "8a12999f8a268c57018a27ebe949151d",
          },
          "Sunday": {
            "id": "8a12999f8a268c57018a27ebe8b4150e",
          },
        },
        "id": "8a12999f8a268c57018a27ebe868150c",
        "pricing": {
          "GBP": 33.99,
        },
      },
    },
  },
  "SubscriptionCard": {
    "billingSystem": "zuora",
    "ratePlans": {
      "Everyday": {
        "billingPeriod": "Month",
        "charges": {
          "Friday": {
            "id": "2c92a00870ec598001710740c91d2f4d",
          },
          "Monday": {
            "id": "2c92a00870ec598001710740c7b82f1c",
          },
          "Saturday": {
            "id": "2c92a00870ec598001710740c8652f37",
          },
          "Sunday": {
            "id": "2c92a00870ec598001710740c9d72f61",
          },
          "Thursday": {
            "id": "2c92a00870ec598001710740c8c42f40",
          },
          "Tuesday": {
            "id": "2c92a00870ec598001710740c80f2f26",
          },
          "Wednesday": {
            "id": "2c92a00870ec598001710740c9802f59",
          },
        },
        "id": "2c92a00870ec598001710740c78d2f13",
        "pricing": {
          "GBP": 69.99,
        },
      },
      "Saturday": {
        "billingPeriod": "Month",
        "charges": {
          "Saturday": {
            "id": "2c92a00870ec598001710740ce042fcb",
          },
        },
        "id": "2c92a00870ec598001710740cdd02fbd",
        "pricing": {
          "GBP": 15.99,
        },
      },
      "Sixday": {
        "billingPeriod": "Month",
        "charges": {
          "Friday": {
            "id": "2c92a00870ec598001710740cb4e2f6b",
          },
          "Monday": {
            "id": "2c92a00870ec598001710740cbb32f77",
          },
          "Saturday": {
            "id": "2c92a00870ec598001710740cd6e2fa2",
          },
          "Thursday": {
            "id": "2c92a00870ec598001710740cc9b2f88",
          },
          "Tuesday": {
            "id": "2c92a00870ec598001710740cc2c2f80",
          },
          "Wednesday": {
            "id": "2c92a00870ec598001710740cd012f90",
          },
        },
        "id": "2c92a00870ec598001710740ca532f69",
        "pricing": {
          "GBP": 61.99,
        },
      },
      "Sunday": {
        "billingPeriod": "Month",
        "charges": {
          "Sunday": {
            "id": "2c92a00870ec598001710740d1103019",
          },
        },
        "id": "2c92a00870ec598001710740d0d83017",
        "pricing": {
          "GBP": 15.99,
        },
      },
      "Weekend": {
        "billingPeriod": "Month",
        "charges": {
          "Saturday": {
            "id": "2c92a00870ec598001710740d28e3024",
          },
          "Sunday": {
            "id": "2c92a00870ec598001710740d325302c",
          },
        },
        "id": "2c92a00870ec598001710740d24b3022",
        "pricing": {
          "GBP": 26.99,
        },
      },
    },
  },
  "SupporterPlus": {
    "billingSystem": "zuora",
    "ratePlans": {
      "Annual": {
        "billingPeriod": "Annual",
        "charges": {
          "Contribution": {
            "id": "8a12892d85fc6df4018602451322287f",
          },
          "Subscription": {
            "id": "8a128ed885fc6ded01860228f7cb3d5f",
          },
        },
        "id": "8a128ed885fc6ded01860228f77e3d5a",
        "pricing": {
          "AUD": 200,
          "CAD": 150,
          "EUR": 120,
          "GBP": 120,
          "NZD": 200,
          "USD": 150,
        },
      },
      "Monthly": {
        "billingPeriod": "Month",
        "charges": {
          "Contribution": {
            "id": "8a128d7085fc6dec01860234cd075270",
          },
          "Subscription": {
            "id": "8a128ed885fc6ded018602296af13eba",
          },
        },
        "id": "8a128ed885fc6ded018602296ace3eb8",
        "pricing": {
          "AUD": 20,
          "CAD": 15,
          "EUR": 12,
          "GBP": 12,
          "NZD": 20,
          "USD": 15,
        },
      },
      "V1DeprecatedAnnual": {
        "billingPeriod": "Annual",
        "charges": {
          "Subscription": {
            "id": "8a12865b8219d9b40182210618c664c1",
          },
        },
        "id": "8a12865b8219d9b40182210618a464ba",
        "pricing": {
          "AUD": 160,
          "CAD": 120,
          "EUR": 95,
          "GBP": 95,
          "NZD": 160,
          "USD": 120,
        },
      },
      "V1DeprecatedMonthly": {
        "billingPeriod": "Month",
        "charges": {
          "Subscription": {
            "id": "8a12865b8219d9b401822106194e64e3",
          },
        },
        "id": "8a12865b8219d9b401822106192b64dc",
        "pricing": {
          "AUD": 17,
          "CAD": 13,
          "EUR": 10,
          "GBP": 10,
          "NZD": 17,
          "USD": 13,
        },
      },
    },
  },
  "TierThree": {
    "billingSystem": "zuora",
    "ratePlans": {
      "DomesticAnnual": {
        "billingPeriod": "Annual",
        "charges": {
          "GuardianWeekly": {
            "id": "8a1288a38ff2af980190025b33191cd7",
          },
          "SupporterPlus": {
            "id": "8a1288a38ff2af980190025b32cb1ccf",
          },
        },
        "id": "8a1288a38ff2af980190025b32591ccc",
        "pricing": {
          "AUD": 680,
          "CAD": 546,
          "EUR": 438,
          "GBP": 300,
          "NZD": 800,
          "USD": 510,
        },
      },
      "DomesticAnnualV2": {
        "billingPeriod": "Annual",
        "charges": {
          "GuardianWeekly": {
            "id": "8a128dfb91f04b9a0191fa30ae951b88",
          },
          "NewspaperArchive": {
            "id": "8a12891291f04b9d0191fa96795b7f22",
          },
          "SupporterPlus": {
            "id": "8a128dfb91f04b9a0191fa30ae611b80",
          },
        },
        "id": "8a128dfb91f04b9a0191fa30ae2e1b7e",
        "pricing": {
          "AUD": 682,
          "CAD": 548,
          "EUR": 440,
          "GBP": 302,
          "NZD": 802,
          "USD": 512,
        },
      },
      "DomesticMonthly": {
        "billingPeriod": "Month",
        "charges": {
          "GuardianWeekly": {
            "id": "8a1299788ff2ec100190025fcd8a2bbb",
          },
          "SupporterPlus": {
            "id": "8a1299788ff2ec100190025fcd232bb3",
          },
        },
        "id": "8a1299788ff2ec100190025fccc32bb1",
        "pricing": {
          "AUD": 60,
          "CAD": 48,
          "EUR": 38.5,
          "GBP": 27,
          "NZD": 70,
          "USD": 45,
        },
      },
      "DomesticMonthlyV2": {
        "billingPeriod": "Month",
        "charges": {
          "GuardianWeekly": {
            "id": "8a128dfb91f04b9a0191fa315d5b1c5b",
          },
          "NewspaperArchive": {
            "id": "8a129c2591f06a5d0191fa945c9177cf",
          },
          "SupporterPlus": {
            "id": "8a128dfb91f04b9a0191fa315d251c53",
          },
        },
        "id": "8a128dfb91f04b9a0191fa315d091c51",
        "pricing": {
          "AUD": 62,
          "CAD": 50,
          "EUR": 40.5,
          "GBP": 29,
          "NZD": 72,
          "USD": 47,
        },
      },
      "RestOfWorldAnnual": {
        "billingPeriod": "Annual",
        "charges": {
          "GuardianWeekly": {
            "id": "8a12894e8ff2af99019002511bdd51ca",
          },
          "SupporterPlus": {
            "id": "8a1281f38ff2af9201900255999049db",
          },
        },
        "id": "8a1299788ff2ec100190024d1e3b1a09",
        "pricing": {
          "GBP": 417.6,
          "USD": 546,
        },
      },
      "RestOfWorldAnnualV2": {
        "billingPeriod": "Annual",
        "charges": {
          "GuardianWeekly": {
            "id": "8a129c2591f06a5d0191fa2edbc63030",
          },
          "NewspaperArchive": {
            "id": "8a12831491f04b9f0191fa9a1f084ec0",
          },
          "SupporterPlus": {
            "id": "8a129c2591f06a5d0191fa2edb803028",
          },
        },
        "id": "8a129c2591f06a5d0191fa2edb383026",
        "pricing": {
          "GBP": 419.6,
          "USD": 548,
        },
      },
      "RestOfWorldMonthly": {
        "billingPeriod": "Month",
        "charges": {
          "GuardianWeekly": {
            "id": "8a128ab18ff2af9301900255d86a79b6",
          },
          "SupporterPlus": {
            "id": "8a128ab18ff2af9301900255d80479ae",
          },
        },
        "id": "8a128ab18ff2af9301900255d77979ac",
        "pricing": {
          "GBP": 36.8,
          "USD": 48,
        },
      },
      "RestOfWorldMonthlyV2": {
        "billingPeriod": "Month",
        "charges": {
          "GuardianWeekly": {
            "id": "8a12891291f04b9d0191fa2ffc46097f",
          },
          "NewspaperArchive": {
            "id": "8a129b7891f06a5a0191fa9888db07de",
          },
          "SupporterPlus": {
            "id": "8a12891291f04b9d0191fa2ffc140977",
          },
        },
        "id": "8a12891291f04b9d0191fa2ffbe10975",
        "pricing": {
          "GBP": 38.8,
          "USD": 50,
        },
      },
    },
  },
}
`;

exports[`Generated product catalog types match snapshot 1`] = `
{
  "Contribution": {
    "billingPeriods": [
      "Annual",
      "Month",
    ],
    "productRatePlans": {
      "Annual": {
        "Contribution": {},
      },
      "Monthly": {
        "Contribution": {},
      },
    },
  },
  "DigitalSubscription": {
    "billingPeriods": [
      "Quarter",
      "Month",
      "Annual",
    ],
    "productRatePlans": {
      "Annual": {
        "Subscription": {},
      },
      "Monthly": {
        "Subscription": {},
      },
      "OneYearGift": {
        "Subscription": {},
      },
      "ThreeMonthGift": {
        "Subscription": {},
      },
    },
  },
  "GuardianLight": {
    "billingPeriods": [
      "Month",
    ],
    "productRatePlans": {
      "Monthly": {
        "GuardianLight": {},
      },
    },
  },
  "GuardianPatron": {
    "billingPeriods": [
      "Month",
    ],
    "currencies": [
      "NZD",
      "CAD",
      "AUD",
      "USD",
      "GBP",
      "EUR",
    ],
    "productRatePlans": {
      "GuardianPatron": {
        "GuardianPatron": {},
      },
    },
  },
  "GuardianWeeklyDomestic": {
    "billingPeriods": [
      "Annual",
      "Quarter",
      "Month",
    ],
    "productRatePlans": {
      "Annual": {
        "Subscription": {},
      },
      "Monthly": {
        "Subscription": {},
      },
      "OneYearGift": {
        "Subscription": {},
      },
      "Quarterly": {
        "Subscription": {},
      },
      "ThreeMonthGift": {
        "Subscription": {},
      },
    },
  },
  "GuardianWeeklyRestOfWorld": {
    "billingPeriods": [
      "Month",
      "Annual",
      "Quarter",
    ],
    "productRatePlans": {
      "Annual": {
        "Subscription": {},
      },
      "Monthly": {
        "Monthly": {},
      },
      "OneYearGift": {
        "Subscription": {},
      },
      "Quarterly": {
        "Subscription": {},
      },
      "ThreeMonthGift": {
        "Subscription": {},
      },
    },
  },
  "HomeDelivery": {
    "billingPeriods": [
      "Month",
    ],
    "productRatePlans": {
      "Everyday": {
        "Friday": {},
        "Monday": {},
        "Saturday": {},
        "Sunday": {},
        "Thursday": {},
        "Tuesday": {},
        "Wednesday": {},
      },
      "Saturday": {
        "Saturday": {},
      },
      "Sixday": {
        "Friday": {},
        "Monday": {},
        "Saturday": {},
        "Thursday": {},
        "Tuesday": {},
        "Wednesday": {},
      },
      "Sunday": {
        "Sunday": {},
      },
      "Weekend": {
        "Saturday": {},
        "Sunday": {},
      },
    },
  },
  "NationalDelivery": {
    "billingPeriods": [
      "Month",
    ],
    "productRatePlans": {
      "Everyday": {
        "Friday": {},
        "Monday": {},
        "Saturday": {},
        "Sunday": {},
        "Thursday": {},
        "Tuesday": {},
        "Wednesday": {},
      },
      "Sixday": {
        "Friday": {},
        "Monday": {},
        "Saturday": {},
        "Thursday": {},
        "Tuesday": {},
        "Wednesday": {},
      },
      "Weekend": {
        "Saturday": {},
        "Sunday": {},
      },
    },
  },
  "SubscriptionCard": {
    "billingPeriods": [
      "Month",
    ],
    "productRatePlans": {
      "Everyday": {
        "Friday": {},
        "Monday": {},
        "Saturday": {},
        "Sunday": {},
        "Thursday": {},
        "Tuesday": {},
        "Wednesday": {},
      },
      "Saturday": {
        "Saturday": {},
      },
      "Sixday": {
        "Friday": {},
        "Monday": {},
        "Saturday": {},
        "Thursday": {},
        "Tuesday": {},
        "Wednesday": {},
      },
      "Sunday": {
        "Sunday": {},
      },
      "Weekend": {
        "Saturday": {},
        "Sunday": {},
      },
    },
  },
  "SupporterPlus": {
    "billingPeriods": [
      "Month",
      "Annual",
    ],
    "productRatePlans": {
      "Annual": {
        "Contribution": {},
        "Subscription": {},
      },
      "Monthly": {
        "Contribution": {},
        "Subscription": {},
      },
      "V1DeprecatedAnnual": {
        "Subscription": {},
      },
      "V1DeprecatedMonthly": {
        "Subscription": {},
      },
    },
  },
  "TierThree": {
    "billingPeriods": [
      "Annual",
      "Month",
    ],
    "productRatePlans": {
      "DomesticAnnual": {
        "GuardianWeekly": {},
        "SupporterPlus": {},
      },
      "DomesticAnnualV2": {
        "GuardianWeekly": {},
        "NewspaperArchive": {},
        "SupporterPlus": {},
      },
      "DomesticMonthly": {
        "GuardianWeekly": {},
        "SupporterPlus": {},
      },
      "DomesticMonthlyV2": {
        "GuardianWeekly": {},
        "NewspaperArchive": {},
        "SupporterPlus": {},
      },
      "RestOfWorldAnnual": {
        "GuardianWeekly": {},
        "SupporterPlus": {},
      },
      "RestOfWorldAnnualV2": {
        "GuardianWeekly": {},
        "NewspaperArchive": {},
        "SupporterPlus": {},
      },
      "RestOfWorldMonthly": {
        "GuardianWeekly": {},
        "SupporterPlus": {},
      },
      "RestOfWorldMonthlyV2": {
        "GuardianWeekly": {},
        "NewspaperArchive": {},
        "SupporterPlus": {},
      },
    },
  },
}
`;
